libsignal-service/src/gv2/api.rs

```diff
@@ -0,0 +1,195 @@
+}
+
+impl<'a, S: PushService, C: CredentialsCache> GroupsV2Api<'a, S, C> {
+    pub fn new(
+        push_service: S,
+        credentials_cache: &'a mut C,
+        server_public_params: ServerPublicParams,
+    ) -> Self {
+        Self {
+            push_service,
+            credentials_cache,
+            server_public_params,
+        }
+    }
+
+    pub async fn get_authorization_for_today(
+        &mut self,
+        uuid: Uuid,
+        group_secret_params: GroupSecretParams,
+    ) -> Result<HttpCredentials, ServiceError> {
+        let today = Self::current_time_days();
+        let auth_credential_response = if let Some(auth_credential_response) =
+            self.credentials_cache.get(&today)?
+        {
+            auth_credential_response
+        } else {
+            let credentials_map = self.get_authorization(today).await?.parse();
+            self.credentials_cache.write(credentials_map)?;
+            self.credentials_cache.get(&today)?.ok_or_else(|| {
+                ServiceError::ResponseError {
+                    reason:
+                        "credentials received did not contain requested day"
+                            .into(),
+                }
+            })?
+        };
+
+        self.get_authorization_string(
+            uuid,
+            group_secret_params,
+            auth_credential_response,
+            today as u32,
+        )
+    }
+
+    async fn get_authorization(
+        &mut self,
+        today: i64,
+    ) -> Result<CredentialResponse, ServiceError> {
+        let today_plus_7_days = today + 7;
+
+        let path =
+            format!("/v1/certificate/group/{}/{}", today, today_plus_7_days);
+
+        self.push_service
+            .get_json(Endpoint::Service, &path, None)
+            .await
+    }
+
+    fn current_time_days() -> i64 {
+        let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
+        let today = chrono::Duration::from_std(now).unwrap();
+        today.num_days()
+    }
+
+    fn get_authorization_string(
+        &self,
+        uuid: Uuid,
+        group_secret_params: GroupSecretParams,
+        credential_response: &AuthCredentialResponse,
+        today: u32,
+    ) -> Result<HttpCredentials, ServiceError> {
+        let auth_credential = self
+            .server_public_params
+            .receive_auth_credential(
+                *uuid.as_bytes(),
+                today,
+                credential_response,
+            )
+            .map_err(|e| {
+                log::error!("zero-knowledge group error: {:?}", e);
+                ServiceError::GroupsV2Error
+            })?;
+
+        let mut random_bytes = [0u8; 32];
+        rand::thread_rng().fill_bytes(&mut random_bytes);
+
+        let auth_credential_presentation = self
+            .server_public_params
+            .create_auth_credential_presentation(
+                random_bytes,
+                group_secret_params,
+                auth_credential,
+            );
```

Comment:
This is something we may want to hand over to a threadpool, but that's of course backend-dependent...

Reply:
At this point, since `actix` is based on `tokio` I'd just use `tokio::spawn_blocking` and `.await` on the handle. But in general, we do operations like this everywhere, don't we? (so this doesn't really apply to only here)

Reply:
Let's postpone until we decide to merge actix and hyper as features into the main crate. That'll make more sense at that point. Actix has its own threadpool, FWIW.